pkg/kubectl/cmd/apply.go

```diff
@@ -585,33 +599,57 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 		return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf("retrieving original configuration from:\n%v\nfor:", obj), source, err)
 	}
 
-	// Create the versioned struct from the type defined in the restmapping
-	// (which is the API version we'll be submitting the patch to)
-	versionedObject, err := scheme.Scheme.New(p.mapping.GroupVersionKind)
 	var patchType types.PatchType
 	var patch []byte
+	var lookupPatchMeta strategicpatch.LookupPatchMeta
+	var schema oapi.Schema
 	createPatchErrFormat := "creating patch with:\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s\nfor:"
-	switch {
-	case runtime.IsNotRegisteredError(err):
-		// fall back to generic JSON merge patch
-		patchType = types.MergePatchType
-		preconditions := []mergepatch.PreconditionFunc{mergepatch.RequireKeyUnchanged("apiVersion"),
-			mergepatch.RequireKeyUnchanged("kind"), mergepatch.RequireMetadataKeyUnchanged("name")}
-		patch, err = jsonmergepatch.CreateThreeWayJSONMergePatch(original, modified, current, preconditions...)
-		if err != nil {
-			if mergepatch.IsPreconditionFailed(err) {
-				return nil, nil, fmt.Errorf("%s", "At least one of apiVersion, kind and name was changed")
-			}
-			return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf(createPatchErrFormat, original, modified, current), source, err)
-		}
-	case err != nil:
-		return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf("getting instance of versioned object for %v:", p.mapping.GroupVersionKind), source, err)
-	case err == nil:
-		// Compute a three way strategic merge patch to send to server.
+	// Try to use openapi first if the openapi spec is available.
+	// Otherwise, fall back to baked-in types.
+	if p.openapiSchema != nil {
+		schema = p.openapiSchema.LookupResource(p.mapping.GroupVersionKind)
```

Comment:
keeping schema and error handling local to this block would make it easier to follow

```go
if p.openapiSchema != nil {
  if schema := p.openapiSchema.LookupResource(p.mapping.GroupVersionKind); schema != nil {
    lookupPatchMeta := strategicpatch.PatchMetaFromOpenAPI{Schema: schema}
    if openapiPatch, err := strategicpatch.CreateThreeWayMergePatch(original, modified, current, lookupPatchMeta, p.overwrite); err != nil {
      fmt.Fprintf(errOut, "warning: error calculating patch from openapi spec: %v\n", err)
    } else {
      patchType = types.StrategicMergePatchType
      patch = openapiPatch
    }
  }
}

if patch == nil {
  // Fallback to compiled-in or unstructured types
  versionedObject, err := scheme.Scheme.New(p.mapping.GroupVersionKind)
  ...
}
```